flattentool/xml_output.py

```diff
@@ -70,16 +71,19 @@ def dict_to_xml(data, tagname, toplevel=True, nsmap=None):
     return el
 
 
-def toxml(data, xml_root_tag):
+def toxml(data, xml_root_tag, xml_schemas=None, root_list_path='iati-activity'):
     nsmap = {
         # This is "bound by definition" - see https://www.w3.org/XML/1998/namespace
         'xml': 'http://www.w3.org/XML/1998/namespace'
     }
     root = dict_to_xml(data, xml_root_tag, nsmap=nsmap)
+    if xml_schemas is not None:
+        schema_dict = IATISchemaWalker(xml_schemas).create_schema_dict(root_list_path)
+        for element in root:
+            sort_iati_element(element, schema_dict)
     comment = ET.Comment('XML generated by flatten-tool')
     root.insert(0, comment)
     if USING_LXML:
         return ET.tostring(root, pretty_print=True)
     else:
-        indent(root)
```

Comment:
This fixes an error on master, but it’s unrelated to this PR.

Reply:
Can there be a note of what error was fixed in the changelog? Thanks

Reply:
Ah, 4513cd34 introduced an error which this fixes.

But it’s only relevant when using stdlib etree which is unsupported anyway, so maybe not worth mentioning?

Reply:
Ah, 4513cd34 introduced a bug which this fixes. The bugfix means using stdlib etree ought to work… But that’s an unsupported feature anyway, so maybe not worth mentioning?

(Sidenote: personally, I’d suggest either supporting stuff or getting rid of it. I’m not terribly keen on making stuff sort of work, but with warnings saying it’s unsupported. This is illustrative of that. The unsupported feature didn’t work due to a bug; this fixes the bug so now the unsupported feature maybe works. Should that go in the changelog? It’s still unsupported, so maybe not…?)

Reply:
I'm thinking of a change log as being useful for 2 things (tho there may be more):

  *  as a user of the software, i am changing from using an older version to the later version and I want to know what has changed so I can make sure my use still works.
  *  as a developer working on the software, i want to be able to work out why the code that is there does what it does (git commit messages should help a bit with this, but more info is good)

None of that depends on whether the feature is supported or not; in either case I still want to know what has changed. So on first think (so to speak) I'm gonna say I think it should be in the changelog. (It can be marked unsupported.)

In the long term, I think I agree with you on either supporting something properly - and making sure it's clearly documented so that new developers coming in have a chance of working out what is going on - or just removing it. Otherwise we have this weird situation where a new developer might do some work which is great with all the supported functions but then they are told "P.R. refused, this breaks a thing you had no chance of knowing about, you fool!". (Not that we tell people they are fools but you can see how this would be frustrating.) Basically, it increases barriers to knowledge sharing.

Reply:
kk – I’ve sent this as a separate PR (#212) and added a changelog entry there.